### ISSUE TITLE
Errors using predict with BayesianNetwork

### ISSUE THREAD

**Issue by TotoRafa** [jmschrei/pomegranate#707]
Hello,
I have some problems using the predict() function for the BayesianNetwork. In fact, I am used to scikit-learn where my approach would be to: 
- Divide the dataset into two differents data set -> training, testing
- retrieve X_train, Y_train, X_test and Y_Test set.
- Use fit function to train the model with X_train and Y_train and take the returned model.
- Use the model to predict using X_test and Y_test 

I was trying to use the same methodology with BayesianNetwork of pomegranate but it does not work. I saw that the simple_samples function should be used instead of fit(). The only way I found how to "make it work" is the following:
`model = BayesianNetwork.from_samples(Alldataset,algorithm='exact')`
`prediction = model.predict(dataset)`
Which I do not really understand the logic... What should I do? How can I have the same approach as scikit-learn?
I am sorry if this is not in the right place or if it has already been answered.
Thanks !

**Comment by jmschrei** [comment]
Howdy

Machine learning models usually have three components: (1) initialization, (2) parameter fitting, and (3) prediction. For many classic machine learning models, such as logistic regression or random forests or the many other ones implemented in scikit-learn, the initialization step is somewhat unimportant or reasonable defaults are known. Thus, the API focuses on steps 2 (through `fit`) and 3 (through `predict`).

In my experience, the initialization of probabilistic models is more complex than classic ML algorithms because the initialization of probabilistic models can involve a lot more structure. To focus on Bayesian networks, "learning" a Bayesian network can mean either learning the structure and parameters of the Bayesian network jointly, or simply updating the parameters given a fixed structure. 

Accordingly, pomegranate has two fitting functions. The `fit` method follows scikit-learn, which is taking an initialized model (such as a Bayesian network with a fixed structure) and learning only the parameters given data. The `from_samples` method wraps the first two steps together by learning the structure of the networks and the parameters of the network at the same time given a data set.

If you want to follow your strategy, I would advise doing `from_samples` on X_train in order to learn your model and then evaluate it using X_test. This is the same procedure you would use for scikit-learn. A difference is that Bayesian networks are unsupervised models, meaning that you do not need to feed in class labels and so y_train and y_test are unnecessary.

**Comment by TotoRafa** [comment]
Thanks for your reply!
First of all, I wanted to tell you that your reply is very completed and I have learned some things ! I also though that from_samples was doing both step one and two!  I tried using X_train for learning and X_test for evaluation but I got some errors... I will check more, it might be because of the structure of my data even though I doubt it.

**Comment by jmschrei** [comment]
I'm going to close this issue but please re-open (or open another issue) with a minimally reproducing example if your code continues to fail. Remember that `predict` for Bayesian networks means filling in an incomplete data matrix, not going from a matrix of data `X` to a label vector `y` like sklearn models.
